Month3/LinkedList/LinkedList/TSLList.h

```diff
@@ -0,0 +1,334 @@
+	}
+	
+
+
+
+	//Size-------------------------------------
+
+	bool IsEmpty()
+	{
+		return Size == 0;
+	}
+
+	int GetSize() 
+	{
+		return Size;
+	}
+
+
+	// Modificators
+
+	void AddHead(const T& NewElement) 
+	{
+		
+		FNode* NewNode = new FNode;
+		NewNode->Element = NewElement;
+		NewNode->Next = Head;
+		Head = NewNode;
+		if (Size == 0) 
+		{
+			Tail = NewNode;
+		}
+		
+		Size++;
+	}
+
+	void AddTail(const T& NewElement) 
+	{
+		if (Tail)
+		{
+			FNode* NewNode = new FNode;
+			NewNode->Element = NewElement;
+			NewNode->Next = nullptr;
+
+			Tail->Next = NewNode;
+			Tail = NewNode;
+
+			Size++;
+		}
+		else 
+		{
+			AddHead(NewElement);
+		}
+	}
+
+	void Insert(const T& NewElement, const int Index )
+	{
+		if (Index < 0 || Index > Size) 
+		{
+			return;
+		}
+		if (Index == 0)
+		{
+			AddHead(NewElement);
+		}
+		else if (Index == Size) 
+		{
+			AddTail(NewElement);
+		}
+		else 
+		{
+			FNode* NewNode = new FNode;
+			NewNode->Element = NewElement;
+			FNode* Current = Head;
+			FNode* Previous = nullptr;
+			for (int CurrentIndex = 0; CurrentIndex < Index; CurrentIndex++)
+			{
+				Previous = Current;
+				Current = Current->Next;
+			}
+
+			NewNode->Next = Current;
+			Previous->Next = NewNode;
+			Size++;
+		}
+		
+		
+	}
+
+	void Remove(const int Index) 
+	{
+		if (Index >= Size || Index < 0) 
+		{
+			return;
+		}
+		FNode* Current = Head;
+		
+		for (int CurrentIndex = 0; CurrentIndex < Index - 1; CurrentIndex++)
+		{
+			Current = Current->Next;
+		}
+		
+		FNode* After = Current->Next;
+		Current->Next = After->Next;
+		delete After;
+		Size--;
+	}
+
+	void Clear() 
+	{
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++) 
+		{
+			FNode* Current = Head;
+			Head = Head->Next;
+			delete Current;
+		}
+		Tail = nullptr;
+		Size = 0;
+	}
+
+
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++)
+		{
+			Predicate(Current->Element);
+			Current = Current->Next;
+		}
+	}
+
+
+	template<typename Pred>
+
+	T& FindByPredicate(const Pred& Predicate) 
+	{
+		FNode* Current = Head;
+		for (int CurrentIndex; CurrentIndex < Size; CurrentIndex++) 
+		{
+			if (Predicate(Current->Element)) 
+			{
+				return *Current;
+			}
+			Current = Current->Next;
+		}
+		return nullptr;
+	}
+
+
+	template<typename Pred>
+	TSLList<T> FilterByPredicate(const Pred& Predicate) 
+	{
+		TSLList<T> NewList;
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++) 
+		{
+			if (Predicate(Current->Element)) 
+			{
+				NewList.AddTail(Current->Element);
+			}
+			Current = Current->Next;
+		}
+		return NewList;
+	}
+	
+	template<typename Pred>
+	void RemoveAllByPredicate(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++) 
+		{
+			T Element = Current->Element;
+			Current = Current->Next;
+			if (Predicate(Current->Element))
```

Comment:
you are not passing the head to the Predicate function the `Current = Current->Next` should be before the if statment